string state
string kind
string level
